support PCD filtering with validEventIds

saurfang opened the pull request:
fixes https://github.com/proofcarryingdata/zupass/issues/744 https://github.com/proofcarryingdata/zupass/issues/789
relates to https://github.com/proofcarryingdata/zupass/issues/470 

The main goal of this PR is to support filtering PCDs in prove screen for ZKEdDSAEventTicketPCD using separate validator params. The key changes are in `pcd.ts`, introduction of `getProveDisplayOptions`, and `<PCDArg />` which consumes the new options.

There are two auxiliary changes explained below due to more strict types.

## type PCDArgs as object type

All PCD arguments are currently typed as `interface`. However `interface` does not extend `Record<>` because they don't have an index signature. In order for us to strongly typecheck Args, we have to migrate all Args to `type` instead.

This is a necessary pre-requisite to ensure `getProveDisplayOptions` can be typed along side PCDPackage's Args.

## SemaphoreIdentityPCDArgs

`SemaphoreIdentityPCD` is the only exception whose prove input is a raw value instead of `Argument<>`. See https://github.com/proofcarryingdata/zupass/issues/830 for more details. To avoid ballooning this PR, `SemaphoreIdentityPCDArgs` is marked with `@ts-expect-error`. 

Note this is strictly due to the above type change because we are more strongly typing `Args` as `extends Record<PropertyKey, Argument<any>>`.  This has no runtime effect.

## getProveDisplayOptions

Similar to `getDisplayOptions` which controls how a PCD is displayed in the passport, `getProveDisplayOptions` controls how prove screen is displayed. It only allows args customization now but can be extended later on screen level customization as found in other specific screens today.

These options are extended over the Argument type and feed into the options as `defaultArgs`. This allows PCD developer to define reasonable display defaults which can be overridden by integrating application with values that is more suitable to the context.

### UI Customization

`displayName` and `description` allow more user friendly name and tooltip for each argument. They also avoid these being passed in the proof get request.

`defaultVisible` allows hide certain advanced arguments by default. For example, in the TG auth bot use case, it is unlikely user would bother checking their numeric Telegram ID. 

### PCD Filtering

`isValid`  provides a way to allow PCD to instruct input validation of the prove arguments. For our purpose, we would like to only display tickets that can be accepted by the proof. Specifically, we implemented the validation over `eventIds, productIds` params. As discussed in the linked design doc, this can be further extended with additional arguments for simple use case or SQL like query text in the future for free form filtering.

## Testing

### Telegram Auth Prove Screen

<img width="406" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/9b6e41f4-6a56-4bbd-8098-ab64e7e1a1e2">

Additional arguments can be shown

<img width="390" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/7dc33fb9-a533-4075-a682-2f28cd27a9a8">

Revealed information is in compact mode by default because they are not changeable by user but can be expanded

<img width="396" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/48f04548-65e8-4b1a-bf23-855faba276b6">

### Event filtering

Using the consumer-client example, cases when no tickets were found that match the event id

<img width="420" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/25d3a438-5f5b-4fe7-803d-6fb07714c0e5">

Using correct event id, corresponding event ticket shows up

<img width="843" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/d5f4b9c8-05b5-452d-8ab6-26e2b3c50a09">

Tested with product ids as well

### Editable toggle list

When toggle list is specified as user provided, the full list is expanded by default and changeable.

<img width="496" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/3cb41798-4eeb-4470-993e-b02ff974a148">

robknight commented:
It's also worth considering filtering on product ID, in cases where we want the user to prove that they have an organizer ticket rather than a general attendance ticket.

ichub commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -1,3 +1,9 @@
+import { useEffect, useState } from "react";
+
```
extra line here - make sure to have the vs code prettier extension installed and that it has formatted all the files in your PR.

ichub commented on apps/consumer-client/src/pages/examples/add-pcd.tsx:
```diff
@@ -375,7 +378,10 @@ async function addSignatureProofPCD(messageToSign: string) {
-    identity: new Identity()
+    identity: {
+      argumentType: ArgumentTypeName.Identity,
+      value: new Identity()
```
good catch. I think one more level of refactor to the prove function of the identity PCD would be good - that it takes an object argument `identity` that is *not* an instance of `Identity`. This would enable it to be used in the same way as the other PCDs. Right now I don't think it's possible to request a proof via the url scheme of an identity PCD (not that you would want to / should do that anyways).

ichub commented on apps/passport-client/components/screens/ProveScreen/GenericProveSection.tsx:
```diff
@@ -39,6 +42,7 @@ export function GenericProveSection<T extends PCDPackage = PCDPackage>({
   const [error, setError] = useState<string | undefined>();
   const [proving, setProving] = useState(false);
   const pcdPackage = pcds.getPackage<T>(pcdType);
+  console.log(pcds, pcdType, pcdPackage);
```
remove this console.log

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -18,22 +19,32 @@ import {
 import { icons } from "../icons";
 
-type ArgSetter = (argName: string, value: any) => void;
+interface ArgInputProps<A extends Argument<any, any>> {
```
I think this deserves a bit of a comment

ichub replied:
would it work with `unknown`s rather than `any`s?

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -18,22 +19,32 @@ import {
+  arg: A;
+  argName: string;
+  setArg: (value: A["value"]) => void;
+  isValid: <A extends Argument<any, any>>(arg: RawValueType<A>) => boolean;
```
this function probably deserves its own type

ichub commented on apps/passport-client/src/appHooks.ts:
```diff
@@ -1,10 +1,13 @@
+import { useContext, useEffect, useState } from "react";
+
```
you've got many files in which you added extra whitespace into the imports section, plz fix that across all the files

ichub commented on apps/passport-client/src/appHooks.ts:
```diff
@@ -29,23 +32,22 @@ export function usePCDCollectionWithHash(): {
 
 export function usePCDs(): PCD[] {
-  const { pcds } = usePCDCollectionWithHash();
+  const pcds = usePCDCollection();
```
I *think* this change could be breaking. There was a reason I had `usePCDCollectionWithHash` here before. If we don't do it this way, there's a chance that changes to the pcd collection would not cause React to re-render. It's been a couple months since I touched the code so I could be mistaken, but I'm pretty sure that was the reason I made it be this way, at least. What was the reason you decided to change it here?

saurfang replied:
Interesting. No particular motivation other than the fact if `  const pcds = usePCDCollection();` depends on the useEffect for it to be up to date, that feels like code smell. I wonder if that is because we mutate `state.pcds` in-place somewhere, otherwise the appSelect should trigger an udpate.
Anyways, I will revert this change since it is irrelevant.

ichub commented on apps/passport-server/src/services/telegramService.ts:
make sure these changes are ok with @cha0sg0d and @lermchair

ichub commented on packages/pcd-types/src/pcd.ts:
```diff
@@ -209,6 +224,7 @@ export enum ArgumentTypeName {
   StringArray = "StringArray",
   PCD = "PCD",
   ToggleList = "ToggleList",
+  Identity = "Identity",
```
All the types here should be trivially `JSON.serialize`-able. Otherwise they would not be able to be robustly serialized and then deserialized, which is what happens to the arguments data structure at the interface between third party applications and zupass.

ichub commented on packages/semaphore-identity-pcd/src/SemaphoreIdentityPCD.ts:
```diff
@@ -1,15 +1,23 @@
-export interface SemaphoreIdentityPCDArgs {
-  identity: Identity;
-}
+export type SemaphoreIdentityPCDArgs = {
+  identity: IdentityArgument;
+};
```
@artwyman related to our discussion about PCD args vs. just a prove function that lets you pass in the right stuff without the args data structure.

rrrliu commented:
@saurfang one note - and I can do this separately after this is merged too - if there are no tickets available in the dropdown, we should show an error like "you do not have a ticket to generate a proof for these events"

saurfang commented:
Addressed all the comments:

* As discussed offline, we are leaving the Identity related change out of this PR. PR description has been updated accordingly with linked the issue. I have spot checked all consumer-client example and went through the reset flow.
* Per @gubsheep suggestion, instead of depending on `validEventids` proof args, the validator now define its own params as a field of the argument. This provides the flexibility for the PCD developer to push validation into the circuit or at the time of consuming PCD.
   * It is notable that this flexibility means one could come up a conflicting pair of proof display constraints and proof constraints (e.g. mismatch `args.ticket.validatorParams.eventIds` and `args.validEventIds`.) However, this is a necessary evil as Brian pointed out that the validator input might not always be available in the proof input (e.g. off-chain group list.)
* To improve the ergonomic, the display related options are added directly to Argument and prove screen options now accepts a list of `defaultArgs` instead. This gives further flexibility to the integrating application and give a piece of mind to PCD developers where their default options can easily be overridden.

ptal? @ichub

ichub commented on apps/consumer-client/src/pages/examples/zk-eddsa-event-ticket-proof.tsx:
```diff
@@ -35,12 +35,26 @@ export default function Page() {
     useState(false);
   const [revealIsConsumed, setRevealIsConsumed] = useState(false);
   const [revealIsRevoked, setRevealIsRevoked] = useState(false);
+  const [revleaFieldsUserProvided, setRevealFieldsUserProvided] =
```
`revlea` -> `reveal`

ichub commented on apps/passport-server/src/services/telegramService.ts:
```diff
@@ -450,12 +450,25 @@ export class TelegramService {
         value: undefined,
-        userProvided: true
+        userProvided: true,
+        displayName: "Your Tikcet",
```
```suggestion
        displayName: "Your Ticket",
```

ichub commented on packages/eddsa-pcd/src/EdDSAPCD.ts:
```diff
@@ -287,10 +285,8 @@ export async function getEdDSAPublicKey(
-      // `F.toObject` converts a point from Montgomery format to a standard one.
-      eddsa.F.toObject(p).toString(16).padStart(64, "0")
-    ) as EdDSAPublicKey;
+  return eddsa.prv2pub(privateKey).map((p) =>
+    // `F.toObject` converts a point from Montgomery format to a standard one.
+    eddsa.F.toObject(p).toString(16).padStart(64, "0")
+  ) as EdDSAPublicKey;
```
just checking but are your prettier settings different from the workspace settings?

saurfang replied:
I don't think so. It only happens to one/two files.

FWIW this is what I from the playground. Can you try save or format this document on your end?

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACATnGBXdVVwAmhAzgIYA6+qNAdAA7oBuATPdgEYAUjAlk2XgBpOAE8AlFRp0AtmXpce41AF4AfFOk0A9NtQADAGK0YEAPIcAVnDAx9qSFCZx0MEqjKp6EXrFQAzdAgZVABZaBgAc2CXUQCIdDkYVFMPVBIYMihCMnRCVGg4Wk0tIlIyWmNTC2tbJRMIAGUYdF9IrgBGADZxBjJCZtyYLi6AFgAaVAoQAAZpyWoaZTJ3AFFCABFGgEEABU4AG14wEVEAbhBxkAh6GF5oEmRQXKCAd13chEeUMgPXslEjyuHHQZDAAGssI0yDI4AAZXxwZD+X4kODA0EQqH0MFtZAtbDokBomS8fHoQlXEhtA5wACK2Ag8GRqKJlhIAA9GjT6YzmUgUQc0VcAI58uC7IL0b4gFYAWigcCIREuIBaZF4RygkQAwsE5MhZQcDqrqdradsYC1eBxsPBdi4EYqWUKiQALGAyA4AdTdvHgJBxYDgjS+-v4-tEhrAJCBICYhIAktkEDBGmBWrdttlmqJaS7hSBGBA0d7QfRDYw4GjmEirr4azBJWRIgaBayrjj0DXDTBRPRqxneLdVXxYN7eIQYG7kAAOGZXTBi3iYZutsgFomZDgTqczpAsK7YNEAFTIHG+gsLcBkHDKRDhWUi2BbcEMCSSd21hrIdogIAAXwAoA

ichub replied:
Nah, it's ok, was just checking.

ichub commented on apps/passport-server/src/services/telegramService.ts:
```diff
@@ -450,12 +450,25 @@ export class TelegramService {
+        displayName: "Your Tikcet",
+        description: "",
+        validatorParams: {
+          eventIds: validEventIds
```
worth adding a `notFoundMessage` here the same way you did in the consumer-client example?

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -45,102 +44,142 @@ type ArgSetter = (argName: string, value: any) => void;
+                argName={key}
+                arg={value as any}
+                setArgs={setArgs}
+                defaultArg={options[key]}
```
```suggestion
                defaultArg={options?.[key]}
```

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -45,102 +44,142 @@ type ArgSetter = (argName: string, value: any) => void;
+          >
+            {showAll ? "▼ Hide" : "► Show"} {hidden.length} more inputs
+          </ShowMoreButton>
+          {showAll &&
```
unless I 'show all', I think the arguments that are hidden never actually get initialized with a value. eg. when i go to consumer client and request a zk ticket pcd proof, and don't expand the hidden items, i am not able to generate a proof because the pcd args did not select my one and only identity. this is a bug.


<img width="456" alt="Screenshot 2023-10-09 at 6 13 30 PM" src="https://github.com/proofcarryingdata/zupass/assets/2636237/9597df7c-1764-474b-8735-7ec33eefe175">

saurfang replied:
oops. good catch

ichub commented on apps/passport-client/components/shared/PCDArgs.tsx:
```diff
@@ -368,16 +388,36 @@ function ToggleListArgInput({
+          onClick={() => setShowAll((showAll) => !showAll)}
+          size="lg"
+        >
+          {showAll ? "-" : "+"}
```
i feel like the icon for showing/hiding all should probably be the same as the one for showing and hiding all the args - rn they're +- and an arrow that turns right or down respectively.

ichub reviewed (commented):
I have found the types here pretty confusing, i feel like it could use one more round of simplification.

saurfang commented on packages/zk-eddsa-event-ticket-pcd/src/ZKEdDSAEventTicketPCD.ts:
```diff
@@ -396,6 +415,78 @@ function claimFromProofResult(
+    defaultArgs: {
+      ticket: {
+        argumentType: ArgumentTypeName.PCD,
+        displayName: "",
```
You are right. This was left in by mistake.

saurfang commented:
updated expand/collapse like so. tested auth flow and verified identity pcd is selected

<img width="411" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/63829e26-5268-4f26-8547-835ad7afb563">
<img width="398" alt="image" src="https://github.com/proofcarryingdata/zupass/assets/4317392/4a948fe9-129c-4c8a-b2f1-5b5d3f94aef8">

ichub reviewed (approved):
nice! lgtm.

*one final thing*

I think the prove screen should not ever leave the loading state in the case that proving succeeded. Right now, when you click prove - it goes into a proving loading state, then when it finishes it turns off the proving state, and briefly flashes the 'prove button', before closing itself. I don't like that flash and we should make it go away. Could you do that in this pr? Should be good to go after.

cha0sg0d commented:
> nice! lgtm.
> 
> _one final thing_
> 
> I think the prove screen should not ever leave the loading state in the case that proving succeeded. Right now, when you click prove - it goes into a proving loading state, then when it finishes it turns off the proving state, and briefly flashes the 'prove button', before closing itself. I don't like that flash and we should make it go away. Could you do that in this pr? Should be good to go after.

+1 thanks for adding this @saurfang!

saurfang commented:
Let me know this looks reasonable. This assumes `onProve` will always handle the UX away from prove screen and all error, especially retry-able ones, will be surfaced. I think that's safe assumption for the time being.

Also fixed an issue on a brief flash of white screen. See following for comparison

[Screen Recording 2023-10-10 at 1.34.31 PM.webm](https://github.com/proofcarryingdata/zupass/assets/4317392/c2aa1986-67ac-4391-8aa5-269ddb99577c)
[Screen Recording 2023-10-10 at 1.41.19 PM.webm](https://github.com/proofcarryingdata/zupass/assets/4317392/6214b826-d03a-412c-8115-3bc0a9da965d)

ichub commented:
> Also fixed an issue on a brief flash of white screen. See following for comparison

nice, looks great!

> Let me know this looks reasonable. This assumes onProve will always handle the UX away from prove screen and all error, especially retry-able ones, will be surfaced. I think that's safe assumption for the time being.

I think that is fine.

🚢 it

saurfang commented:
:shipit: 🚢 🤞
